pxr/base/tf/delegatedCountPtr.h

```diff
@@ -0,0 +1,268 @@
+/// and `TfDelegatedCountDecrement` to bookkeep.
+///
+/// The `TfDelegatedCountPtr` is responsible for calling
+/// `TfDelegatedCountIncrement` and `TfDelegatedCountDecrement` as needed in 
+/// construction, assignment, and destruction operations.
+///
+/// `TfDelegatedCountIncrement` and `TfDelegatedCountDecrement` are expected to
+/// be `noexcept`. The increment and decrement methods can assume pointers are
+/// non-null. Freeing of memory is delegated to `TfDelegatedCountDecrement`.
+///
+/// This may be created from a user constructed raw pointer or via
+/// `TfMakeDelegatedCountPtr`.
+template <typename ValueType>
+class TfDelegatedCountPtr {
+public:
+    using RawPtrType = std::add_pointer_t<ValueType>;
+    using ReferenceType = std::add_lvalue_reference_t<ValueType>;
+    using element_type = ValueType;
+
+private:
+    template <typename ConvertibleType>
+    using _IsPtrConvertible = std::is_convertible<
+        std::add_pointer_t<ConvertibleType>, RawPtrType>;
+
+public:
+    /// Create a pointer storing `nullptr`
+    TfDelegatedCountPtr() noexcept = default;
+
+    /// Creates a new pointer without retaining
+    /// \sa TfDelegatedCountDoNotIncrementTag
+    TfDelegatedCountPtr(TfDelegatedCountDoNotIncrementTagType,
+                        RawPtrType rawPointer) noexcept :
+        _pointer{rawPointer} {
+    }
+
+    /// Creates a new retained pointer
+    ///
+    /// Retains only if the pointer is valid.
+    /// \sa TfDelegatedCountIncrementTag
+    TfDelegatedCountPtr(TfDelegatedCountIncrementTagType,
+                        RawPtrType rawPointer) noexcept :
+        _pointer{rawPointer} {
+        _RetainIfValid();
+    }
+
+    /// Constructs by copying the pointer from `ptr`.
+    ///
+    /// Retains only if the pointer is valid.
+    TfDelegatedCountPtr(const TfDelegatedCountPtr& ptr) noexcept :
+        _pointer{ptr.get()} {
+        _RetainIfValid();
+    }
+
+    /// Constructs by copying a convertible pointer from `ptr`.
+    ///
+    /// This is necessary in part to allow a `const` `ValueType` to be
+    /// constructed from from a non-`const` `ValueType`.
+    ///
+    /// Retains only if the pointer is valid.
+    template <typename OtherType>
+    explicit TfDelegatedCountPtr(
+        const TfDelegatedCountPtr<OtherType>& ptr,
+        std::enable_if_t<_IsPtrConvertible<OtherType>::value, int> = 0) noexcept :
+        _pointer(ptr.get()) {
+        _RetainIfValid();
+    }
+
+    /// Constructs by moving the pointer from `ptr`.
+    ///
+    /// `ptr` will be reset to its default state (ie. `nullptr`).
+    TfDelegatedCountPtr(TfDelegatedCountPtr&& ptr) noexcept :
+        _pointer(ptr.get()) {
+        ptr._pointer = nullptr;
+    }
+
+    /// Assigns by copying the pointer from `ptr`.
+    ///
+    /// Retains only if the pointer is valid.
+    TfDelegatedCountPtr& operator=(const TfDelegatedCountPtr& ptr) noexcept {
+        if (std::addressof(ptr) != this) {
+            _ReleaseIfValid();
```

Comment:
I think there's a bug here where if `*this` transitively owns `ptr`, then releasing `this` first can cause `ptr` to dangle.  I think we have to retain `ptr` first, then release `*this`.